internal/tofu/context_functions.go

```diff
@@ -3,55 +3,132 @@ package tofu
 import (
 	"errors"
 	"fmt"
-	"log"
-	"sync"
 
+	"github.com/hashicorp/hcl/v2"
 	"github.com/opentofu/opentofu/internal/addrs"
+	"github.com/opentofu/opentofu/internal/configs"
 	"github.com/opentofu/opentofu/internal/providers"
+	"github.com/opentofu/opentofu/internal/tfdiags"
 	"github.com/zclconf/go-cty/cty"
 	"github.com/zclconf/go-cty/cty/function"
 )
 
-// Lazily creates a single instance of a provider for repeated use.
-// Concurrency safe
-func lazyProviderInstance(addr addrs.Provider, factory providers.Factory) providers.Factory {
-	var provider providers.Interface
-	var providerLock sync.Mutex
-	var err error
+// This builds a provider function using an EvalContext and some additional information
+// This is split out of BuiltinEvalContext for testing
+func evalContextProviderFunction(ctx EvalContext, mc *configs.Config, op walkOperation, pf addrs.ProviderFunction, rng tfdiags.SourceRange) (*function.Function, tfdiags.Diagnostics) {
+	var diags tfdiags.Diagnostics
 
-	return func() (providers.Interface, error) {
-		providerLock.Lock()
-		defer providerLock.Unlock()
+	pr, ok := mc.Module.ProviderRequirements.RequiredProviders[pf.Name]
+	if !ok {
+		return nil, diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Unknown function provider",
+			Detail:   fmt.Sprintf("Provider %q does not exist within the required_providers of this module", pf.Name),
+			Subject:  rng.ToHCL().Ptr(),
+		})
+	}
+
+	// Very similar to transform_provider.go
+	absPc := addrs.AbsProviderConfig{
+		Provider: pr.Type,
+		Module:   mc.Path,
+		Alias:    pf.Alias,
+	}
+
+	provider := ctx.Provider(absPc)
 
+	if provider == nil {
+		// Configured provider (NodeApplyableProvider) not required via transform_provider.go.  Instead we should use the unconfigured instance (NodeEvalableProvider) in the root.
+
+		// Make sure the alias is valid
+		validAlias := pf.Alias == ""
+		if !validAlias {
+			for _, alias := range pr.Aliases {
+				println(alias.Alias)
+				if alias.Alias == pf.Alias {
+					validAlias = true
+					break
+				}
+			}
+			if !validAlias {
+				return nil, diags.Append(&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Unknown function provider",
+					Detail:   fmt.Sprintf("Provider %q does not have alias %q", pf.Name, pf.Alias),
```

Comment:
Nit: No provider %q instance with alias %q

or alternatively

does not have **an instace with** alias